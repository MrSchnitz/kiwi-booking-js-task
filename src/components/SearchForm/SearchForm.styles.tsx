import styled from "styled-components";

export const SearchFormWrapper = styled.div`
  padding: 24px;
`;

export const SearchFormPopoverContentWrapper = styled.div`
    max-height: 300px;
    overflow-y: auto;
    
    &::-webkit-scrollbar {
        width: 5px;
    }

    &::-webkit-scrollbar-track {
        background: #ffffff;
    }

    &::-webkit-scrollbar-thumb {
        background: #00A991;
    }

    &::-webkit-scrollbar-thumb:hover {
        background: #007F6D;
    }
`;
