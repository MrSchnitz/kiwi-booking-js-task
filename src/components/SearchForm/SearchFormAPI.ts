import { createSelector, createSlice, PayloadAction } from "@reduxjs/toolkit";
import { all, call, debounce, put, takeEvery } from "redux-saga/effects";
import { RootState } from "../../internals/rootState";
import { SagaIterator } from "redux-saga";
import {
  FlightLocation,
  FlightLocationsResponse,
} from "../../types/FlightLocationTypes";

export enum SearchParam {
  ORIGIN = "origin",
  DESTINATION = "destination",
  DATE_FROM = "dateFrom",
  DATE_TO = "dateTo",
}

export interface ISearchFormData {
  [SearchParam.ORIGIN]: string;
  [SearchParam.DESTINATION]: string;
  [SearchParam.DATE_FROM]: string;
  [SearchParam.DATE_TO]: string;
}

/**
 * API State interface
 */
export interface ISearchFormApi {
  originSearchPhrase: string;
  destinationSearchPhrase: string;
  locations: Array<FlightLocation>;
  searchFormData: ISearchFormData;
}

/**
 * API
 */
class SearchFormApi {
  private static instance: SearchFormApi;

  private constructor() {
    this.handleSearch = this.handleSearch.bind(this);
    this.handleLoadSearchData = this.handleLoadSearchData.bind(this);
    this.saga = this.saga.bind(this);
  }

  public static getInstance(): SearchFormApi {
    if (SearchFormApi.instance) {
      return this.instance;
    }
    this.instance = new SearchFormApi();
    return this.instance;
  }

  /*
   * SLICE
   */
  public getInitialState(): ISearchFormApi {
    return {
      originSearchPhrase: "",
      destinationSearchPhrase: "",
      locations: [],
      searchFormData: {
        origin: "",
        destination: "",
        dateFrom: "",
        dateTo: "",
      },
    };
  }

  public slice = createSlice({
    name: "homePageApiSlice",
    initialState: this.getInitialState(),
    reducers: {
      changeOriginPhrase(state, action: PayloadAction<string>) {
        state.originSearchPhrase = action.payload;
      },
      changeDestinationPhrase(state, action: PayloadAction<string>) {
        state.destinationSearchPhrase = action.payload;
      },
      loadSearchFormData(_, action: PayloadAction<ISearchFormData>) {},
      setLocations(state, action: PayloadAction<Array<FlightLocation>>) {
        state.locations = action.payload;
      },
      setSearchFormData(state, action: PayloadAction<ISearchFormData>) {
        state.searchFormData = action.payload;
      },
    },
  });

  /*
   * SAGAS
   */
  public *handleSearch(action: PayloadAction<string>): SagaIterator {
    const response = yield call(
      fetch,
      `https://api.skypicker.com/locations?term=${action.payload}&location_types=airport`
    );
    const responseBody: FlightLocationsResponse = yield response.json();
    yield put(this.slice.actions.setLocations(responseBody.locations));
  }

  public *handleLoadSearchData({
    payload,
  }: PayloadAction<ISearchFormData>): Generator {
    const searchData: ISearchFormData = { ...payload };
    if (!!searchData[SearchParam.ORIGIN]) {
      const response = yield call(
        this.fetchLocationById,
        searchData[SearchParam.ORIGIN]
      );
      const location: FlightLocation = (response as FlightLocationsResponse)
        .locations[0];
      searchData[SearchParam.ORIGIN] = location.name;
    }
    if (!!searchData[SearchParam.DESTINATION]) {
      const response = yield call(
        this.fetchLocationById,
        searchData[SearchParam.DESTINATION]
      );
      const location: FlightLocation = (response as FlightLocationsResponse)
        .locations[0];
      searchData[SearchParam.DESTINATION] = location.name;
    }

    yield put(this.slice.actions.setSearchFormData(searchData));
  }

  private *fetchLocationById(locationId: string): SagaIterator {
    const response = yield call(
      fetch,
      `https://api.skypicker.com/locations?type=id&id=${locationId}`
    );
    return yield response.json();
  }

  /*
   * SAGA - MAIN
   */
  public *saga(): SagaIterator {
    const { changeOriginPhrase, changeDestinationPhrase, loadSearchFormData } =
      this.slice.actions;
    yield all([
      yield debounce(
        500,
        [changeOriginPhrase.type, changeDestinationPhrase.type],
        this.handleSearch
      ),
      yield takeEvery(loadSearchFormData.type, this.handleLoadSearchData),
    ]);
  }

  /*
   * SELECTORS
   */
  private selectDomain(state: RootState): ISearchFormApi {
    return state.searchFormApiSlice || this.getInitialState();
  }

  public originSearchPhraseSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.originSearchPhrase
  );

  public destinationSearchPhraseSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.destinationSearchPhrase
  );

  public locationsSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.locations
  );

  public searchFormDataSelector = createSelector(
    [this.selectDomain],
    (apiState) => apiState.searchFormData
  );
}

export default SearchFormApi.getInstance();

export const {
  actions: SearchFormAPI,
  reducer: SearchFormApiReducer,
  name: SearchFormApiName,
} = SearchFormApi.getInstance().slice;

export const {
  originSearchPhraseSelector,
  destinationSearchPhraseSelector,
  locationsSelector,
  searchFormDataSelector,
  saga: SearchFormApiSaga,
} = SearchFormApi.getInstance();
