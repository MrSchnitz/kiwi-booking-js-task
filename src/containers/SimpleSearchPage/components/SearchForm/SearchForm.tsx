import { ChangeEvent, useState } from "react";
import AutoCompleteForm from "../AutoComplete/AutoCompleteForm";

export function SearchForm(props: any) {
  const [originValue, setOriginValue] = useState("");
  const [destinationValue, setDestinationValue] = useState("");
  const [departureValue, setDepartureValue] = useState("");

  const handleChangeDepartureDate = (event: ChangeEvent<any>) => {
    setDepartureValue(event.target.value);
  };

  const handleSearch = () => {
    props.onSearch({
      origin: originValue,
      destination: destinationValue,
      departure: departureValue,
    });
  };

  const handleSelectOrigin = (item) => {
    setOriginValue(item);
  };

  const handleSelectDestination = (item) => {
    setDestinationValue(item);
  };

  return (
    <div>
      {originValue ? (
        <div>{originValue}</div>
      ) : (
        <AutoCompleteForm onSelect={handleSelectOrigin} />
      )}
      <AutoCompleteForm onSelect={handleSelectDestination} />
      <input
        type="date"
        onChange={handleChangeDepartureDate}
        value={departureValue}
      />
      <button onClick={handleSearch} disabled={props.isDisabled}>
        Search
      </button>
    </div>
  );
}
