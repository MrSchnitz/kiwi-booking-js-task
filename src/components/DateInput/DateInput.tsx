import React, { useEffect, useState } from "react";
import { InputField } from "@kiwicom/orbit-components";
import ReactDatePicker from "react-datepicker";
import customParseFormat from "dayjs/plugin/customParseFormat";
import dayjs from "dayjs";
import { SearchParam } from "../SearchForm/SearchFormAPI";

dayjs.extend(customParseFormat);

const DATE_FORMAT = "DD/MM/YYYY";

interface IProps {
  type: SearchParam.DATE_FROM | SearchParam.DATE_TO;
  value: string;
  prefix: string;
  onChange(value: string, type: SearchParam): void;
}

const DateInput = React.memo(({ value, prefix, type, onChange }: IProps) => {
  const [date, setDate] = useState<Date | null>(
    !!value ? dayjs(value, DATE_FORMAT).toDate() : null
  );

  useEffect(() => {
    setDate(!!value ? dayjs(value, DATE_FORMAT).toDate() : null);
  }, [value]);

  return (
    <ReactDatePicker
      selected={date}
      placeholderText="Set date..."
      onChange={onInputChange}
      dateFormat="dd/MM/yyyy"
      customInput={
        <InputField prefix={prefix} value={value} placeholder="Set date..." />
      }
    />
  );

  function onInputChange(date: Date | null): void {
    if (dayjs().isBefore(date, "day") || dayjs().isSame(date, "day")) {
      setDate(date);
      onChange(dayjs(date).format(DATE_FORMAT).toString(), type);
    }
  }
});

export default DateInput;
