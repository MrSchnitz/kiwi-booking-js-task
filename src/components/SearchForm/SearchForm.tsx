import React, { useCallback, useEffect, useState } from "react";
import {
  createSearchParams,
  useNavigate,
  useSearchParams,
} from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { Button, Heading, Stack } from "@kiwicom/orbit-components";
import {
  destinationSearchPhraseSelector,
  ISearchFormData,
  locationsSelector,
  originSearchPhraseSelector,
  SearchFormAPI,
  searchFormDataSelector,
  SearchParam,
} from "./SearchFormAPI";
import { SearchFormWrapper } from "./SearchForm.styles";
import { City, FlightLocation } from "./SearchForm.types";
import SearchAutoCompleteInput from "../SearchAutoCompleteInput/SearchAutoCompleteInput";
import DateInput from "../DateInput/DateInput";

interface IProps {
  onSearch?(searchParams: ISearchFormData): void;
}

const SearchForm = React.memo(({ onSearch }: IProps) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [activeInput, setActiveInput] = useState<null | SearchParam>(null);
  const [searchParams, setSearchParams] = useSearchParams();

  const originSearchPhrase: string = useSelector(originSearchPhraseSelector);
  const destinationSearchPhrase: string = useSelector(
    destinationSearchPhraseSelector
  );
  const locations: Array<FlightLocation> = useSelector(locationsSelector);
  const searchFormData: ISearchFormData = useSelector(searchFormDataSelector);

  const isSearchDisabled =
    Array.from(searchParams.values()).filter(Boolean).length !== 4;

  useEffect(() => {
    if (Array.from(searchParams.values()).filter(Boolean).length) {
      const searchData: ISearchFormData = {
        [SearchParam.ORIGIN]: "",
        [SearchParam.DESTINATION]: "",
        [SearchParam.DATE_FROM]: "",
        [SearchParam.DATE_TO]: "",
      };
      const searchDataKeys = Object.keys(searchData);

      for (const [k, v] of searchParams.entries()) {
        if (searchDataKeys.includes(k) && !!v) {
          searchData[k as SearchParam] = v;
        }
      }

      dispatch(SearchFormAPI.loadSearchFormData(searchData));
    }
  }, [dispatch, searchParams]);

  const onDateInputChange = useCallback(
    (dateValue: string, param: SearchParam): void => {
      setSearchParams({
        ...Object.fromEntries(searchParams.entries()),
        [param]: dateValue,
      });
      dispatch(
        SearchFormAPI.setSearchFormData({
          ...searchFormData,
          [param]: dateValue,
        })
      );
    },
    [dispatch, searchFormData, searchParams, setSearchParams]
  );

  const onLocationInputChange = useCallback(
    (value: string, param: SearchParam): void => {
      console.log("AAA", value, param);
      setActiveInput(param);
      dispatch(
        param === SearchParam.ORIGIN
          ? SearchFormAPI.changeFromPhrase(value)
          : SearchFormAPI.changeToPhrase(value)
      );
    },
    [dispatch]
  );

  const onSelectLocation = useCallback(
    (city: City): void => {
      if (activeInput) {
        dispatch(SearchFormAPI.setLocations([]));
        dispatch(
          SearchFormAPI.setSearchFormData({
            ...searchFormData,
            [activeInput]: city.name,
          })
        );
        setSearchParams({
          ...Object.fromEntries(searchParams.entries()),
          [activeInput]: city.id,
        });
        setActiveInput(null);
      }
    },
    [dispatch, activeInput, searchParams, searchFormData, setSearchParams]
  );

  const onSelectedCityButtonClick = useCallback(
    (param: SearchParam): void => {
      const searchParamsCopy = Object.fromEntries(searchParams.entries());
      delete searchParamsCopy[param];
      setSearchParams(searchParamsCopy);
    },
    [searchParams, setSearchParams]
  );

  const onSearchClick = (): void => {
    const searchParamsCopy = Object.fromEntries(searchParams.entries());
    const searchData: ISearchFormData = {
      [SearchParam.ORIGIN]: searchParamsCopy[SearchParam.ORIGIN],
      [SearchParam.DESTINATION]: searchParamsCopy[SearchParam.DESTINATION],
      [SearchParam.DATE_FROM]: searchParamsCopy[SearchParam.DATE_FROM],
      [SearchParam.DATE_TO]: searchParamsCopy[SearchParam.DATE_TO],
    };

    if (onSearch) {
      onSearch(searchData);
      return;
    }

    navigate({
      pathname: "search",
      search: `?${createSearchParams({ ...searchData })}`,
    });
  };

  return (
    <SearchFormWrapper>
      <Heading type={"title2"} spaceAfter="medium">
        Search for flights
      </Heading>
      <Stack
        flex={true}
        direction="column"
        justify="center"
        align="center"
        desktop={{ direction: "row" }}
      >
        <Stack>
          <SearchAutoCompleteInput
            type={SearchParam.ORIGIN}
            activeInput={activeInput}
            locations={locations}
            searchFormData={searchFormData}
            value={originSearchPhrase}
            onLocationInputChange={onLocationInputChange}
            onSelectLocation={onSelectLocation}
            onSelectedCityButtonClick={onSelectedCityButtonClick}
          />
        </Stack>
        <Stack>
          <SearchAutoCompleteInput
            type={SearchParam.DESTINATION}
            activeInput={activeInput}
            locations={locations}
            searchFormData={searchFormData}
            value={destinationSearchPhrase}
            onLocationInputChange={onLocationInputChange}
            onSelectLocation={onSelectLocation}
            onSelectedCityButtonClick={onSelectedCityButtonClick}
          />
        </Stack>
        <DateInput
          type={SearchParam.DATE_FROM}
          value={searchFormData[SearchParam.DATE_FROM]}
          prefix="Departure"
          onChange={onDateInputChange}
        />
        <DateInput
          type={SearchParam.DATE_TO}
          value={searchFormData[SearchParam.DATE_TO]}
          prefix="Return"
          onChange={onDateInputChange}
        />
        <Button
          type="primary"
          disabled={isSearchDisabled}
          onClick={onSearchClick}
        >
          Search
        </Button>
      </Stack>
    </SearchFormWrapper>
  );
});

export default SearchForm;
