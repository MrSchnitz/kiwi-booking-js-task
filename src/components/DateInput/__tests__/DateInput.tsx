import { render } from "@testing-library/react";
import DateInput from "../DateInput";
import { SearchParam } from "../../SearchForm/SearchFormAPI";

describe("DateInput", () => {
  it("should render basic component", () => {
    const { container } = render(
      <DateInput
        type={SearchParam.DATE_FROM}
        value={""}
        prefix={"From"}
        onChange={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();
  });

  it("should rerender on value change", () => {
    const { container, rerender } = render(
      <DateInput
        type={SearchParam.DATE_FROM}
        value={"22/07/2022"}
        prefix={"From"}
        onChange={jest.fn()}
      />
    );

    expect(container).toMatchSnapshot();

    rerender(
      <DateInput
        type={SearchParam.DATE_FROM}
        value={"30/08/2023"}
        prefix={"From"}
        onChange={jest.fn()}
      />
    );
    expect(container).toMatchSnapshot();
  });
});
