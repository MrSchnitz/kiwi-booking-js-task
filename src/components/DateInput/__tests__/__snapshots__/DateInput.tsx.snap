// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DateInput should render basic component 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="InputField__Field-sc-jjv066-0 koAYkz"
      >
        <div
          class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
        >
          <div
            class="InputField__Prefix-sc-jjv066-4 kUJKxj"
          >
            From
          </div>
          <input
            aria-labelledby="1"
            class="InputField__Input-sc-jjv066-6 kPMZTv"
            data-state="ok"
            id="1"
            placeholder="Set date..."
            type="text"
            value=""
          />
          <div
            class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput should rerender on value change 1`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="InputField__Field-sc-jjv066-0 koAYkz"
      >
        <div
          class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
        >
          <div
            class="InputField__Prefix-sc-jjv066-4 kUJKxj"
          >
            From
          </div>
          <input
            aria-labelledby="2"
            class="InputField__Input-sc-jjv066-6 kPMZTv"
            data-state="ok"
            id="2"
            placeholder="Set date..."
            type="text"
            value="22/07/2022"
          />
          <div
            class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`DateInput should rerender on value change 2`] = `
<div>
  <div
    class="react-datepicker-wrapper"
  >
    <div
      class="react-datepicker__input-container"
    >
      <div
        class="InputField__Field-sc-jjv066-0 koAYkz"
      >
        <div
          class="InputField__InputContainer-sc-jjv066-2 bFmIrl"
        >
          <div
            class="InputField__Prefix-sc-jjv066-4 kUJKxj"
          >
            From
          </div>
          <input
            aria-labelledby="2"
            class="InputField__Input-sc-jjv066-6 kPMZTv"
            data-state="ok"
            id="2"
            placeholder="Set date..."
            type="text"
            value="30/08/2023"
          />
          <div
            class="InputField__FakeInput-sc-jjv066-1 dmsAbX"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
